### ISSUE TITLE
Update on batch of single reward episodes

### ISSUE THREAD

**Issue by Deathn0t** [reinforceio/tensorforce#501]
Hi,

Let's say you have only one reward per episode that you can compute once you have a whole episode. Let's say that this function is the `environment.execute(...)`. Depending on the episode the computation of this final reward is taking a different duration. I would like to (idea of the algorithm)👍
```
# Initialisation
generate_episodes(100 episodes)
while True:
  wait_for_10_rewards()
  batch_update_on_10_rewards()
  generate_episodes(10 episodes)
```

I think that each time the agent is acting `self.agent.act(state)`, we are putting in buffers at least the state and the corresponding actions. So if I want to update with `self.agent.observe(...)` in a different order than the one for generating the actions. I think it'll be problematic with the actual implementation, am I right ?

**Comment by AlexKuhnle** [comment]
Yes, mixing `act(..)` and `observe(..)` is currently problematic, the agent is supposed to be single-threaded.  There is no good way to prevent this without changing bits in the code. Since it sounds related to the other issue, see my comment on Gitter and let's discuss a solution there.

**Comment by AlexKuhnle** [comment]
Solved via #508.
